// @hephaestus/utils@1.0.0-alpha.8/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ExtendedMap: () => ExtendedMap,
  identity: () => identity,
  logger: () => logger_exports,
  unknownHasKey: () => unknownHasKey
});
module.exports = __toCommonJS(src_exports);

// src/extended-map.ts
var ExtendedMap = class extends Map {
  find(func) {
    for (const item of this.values()) {
      if (func(item)) {
        return item;
      }
    }
    return void 0;
  }
  filter(func) {
    const arr = [];
    for (const item of this.values()) {
      if (func(item)) {
        arr.push(item);
      }
    }
    return arr;
  }
  map(func) {
    const arr = [];
    for (const item of this.values()) {
      arr.push(func(item));
    }
    return arr;
  }
  reduce(func, initialValue) {
    const iter = this.values();
    let current = iter.next();
    let result = initialValue != null ? initialValue : current.value;
    while (!current.done) {
      result = func(result, current.value);
      current = iter.next();
    }
    return result;
  }
  every(func) {
    for (const item of this.values()) {
      if (!func(item)) {
        return false;
      }
    }
    return true;
  }
  some(func) {
    for (const item of this.values()) {
      if (func(item)) {
        return true;
      }
    }
    return false;
  }
};

// src/identity.ts
function identity(list) {
  return list.filter((x) => x != null);
}

// src/logger.ts
var logger_exports = {};
__export(logger_exports, {
  LEVEL: () => LEVEL,
  error: () => error,
  info: () => info,
  log: () => log,
  success: () => success,
  warn: () => warn
});
var import_util = require("util");
var LEVEL = /* @__PURE__ */ ((LEVEL2) => {
  LEVEL2["SUCCESS"] = "green";
  LEVEL2["WARN"] = "yellow";
  LEVEL2["ERROR"] = "red";
  LEVEL2["INFO"] = "cyan";
  LEVEL2["DEFAULT"] = "white";
  return LEVEL2;
})(LEVEL || {});
function log(content, level = "white" /* DEFAULT */, stream = process.stdout) {
  const codes = import_util.inspect.colors[level];
  const text = content.join(" ");
  const message = codes ? `\x1B[${codes[0]}m${text}\x1B[${codes[1]}m` : text;
  stream.write(Buffer.from(`ERIS_BOILER: ${(0, import_util.formatWithOptions)({ colors: true }, message)}
`));
}
function error(...content) {
  return log(content, "red" /* ERROR */, process.stderr);
}
function info(...content) {
  return log(content, "cyan" /* INFO */);
}
function success(...content) {
  return log(content, "green" /* SUCCESS */);
}
function warn(...content) {
  return log(content, "yellow" /* WARN */);
}

// src/unknown-has-key.ts
function unknownHasKey(value, key) {
  return value != null && typeof value === "object" && key in value;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExtendedMap,
  identity,
  logger,
  unknownHasKey
});
