// discord-player-youtubei@1.4.2/dist/experimental/index.mjs
import {
  extractVideoId
} from "../chunk-FBJVWABG.mjs";

// lib/experimental/PoToken/PoToken.ts
import { BG } from "bgutils-js";
import { Window } from "happy-dom";
var HARD_CODED_REQ_KEY = "O43z0dpjhgX20SCx4KAo";
async function poTokenExtraction(innertube) {
  const visitorData = innertube.session.context.client.visitorData;
  if (!visitorData)
    throw new Error("Innertube instance does not contain visitor data");
  const window = new Window();
  Object.assign(globalThis, {
    document: window.document,
    window
  });
  const bgConfig = {
    // excuse my variable naming
    fetch: (i, a) => fetch(i, a),
    globalObj: globalThis,
    identifier: visitorData,
    requestKey: HARD_CODED_REQ_KEY
  };
  const challenge = await BG.Challenge.create(bgConfig);
  if (!challenge)
    throw new Error("Unable to retrieve challenge data from botguard");
  const jsInter = challenge.interpreterJavascript.privateDoNotAccessOrElseSafeScriptWrappedValue;
  if (jsInter) {
    new Function(jsInter)();
  } else throw new Error("Unable to load botguard's VM");
  return BG.PoToken.generate({
    bgConfig,
    program: challenge.program,
    globalName: challenge.globalName
  });
}

// lib/experimental/ServerAbrStream/index.ts
import { PassThrough } from "stream";
async function getGoogleVideoOrThrow() {
  try {
    return (await import("googlevideo")).default;
  } catch {
    throw new Error(
      'Unable to find googlevideo. Please install it via "npm install googlevideo"'
    );
  }
}
async function createServerAbrStream(track, ext, onError) {
  const innertube = ext.innerTube;
  const { ServerAbrStream } = await getGoogleVideoOrThrow();
  if (!innertube.session.player)
    throw new Error("ServerAbrStream does not work without a valid player.");
  const videoInfo = await innertube.getBasicInfo(
    extractVideoId(track.url),
    "WEB"
  );
  const fmt = videoInfo.chooseFormat({ type: "audio", quality: "best" });
  const audio = {
    itag: fmt.itag,
    lastModified: fmt.last_modified_ms,
    xtags: fmt.xtags
  };
  const sabrStreamUrl = innertube.session.player.decipher(
    videoInfo.page[0].streaming_data?.server_abr_streaming_url
  );
  const videoPlaybackUstreamerConfig = videoInfo.page[0].player_config?.media_common_config.media_ustreamer_request_config?.video_playback_ustreamer_config;
  if (!videoPlaybackUstreamerConfig)
    throw new Error("ustreamerConfig not found");
  if (!sabrStreamUrl) throw new Error("serverAbrStreamingUrl not found");
  const sabrStream = new ServerAbrStream({
    fetch: innertube.session.http.fetch_function,
    serverAbrStreamingUrl: sabrStreamUrl,
    videoPlaybackUstreamerConfig,
    durationMs: (videoInfo.basic_info.duration ?? 0) * 1e3
  });
  const readable = new PassThrough();
  sabrStream.on("data", (data) => {
    for (const formatData of data.initializedFormats) {
      if (formatData.mimeType?.includes("video")) continue;
      const media = formatData.mediaChunks;
      for (const chunk of media) {
        readable.write(chunk);
      }
    }
  });
  sabrStream.on("error", (err) => {
    if (onError) onError(err);
  });
  sabrStream.on("end", () => {
    readable.end();
  });
  sabrStream.init({
    audioFormats: [audio],
    videoFormats: [],
    clientAbrState: {
      startTimeMs: 0,
      mediaType: 1
    }
  });
  return readable;
}

// lib/experimental/index.ts
console.log(
  "\x1B[2;33mYou are using discord-player-youtubei's experimental features. Expect bugs.\x1B[0m"
);
export {
  createServerAbrStream,
  getGoogleVideoOrThrow,
  poTokenExtraction
};
